File: src/api/cost.ts
import { invoke } from '@tauri-apps/api/core'

export interface RequestLog {
  id: string
  provider_id: string | null
  model: string | null
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_creation_tokens: number
  total_cost_cny: number
  latency_ms: number | null
  first_token_ms: number | null
  status_code: number | null
  error_message: string | null
  session_id: string | null
  is_streaming: boolean
  created_at: number
}

export interface DailyUsage {
  date: string
  provider_id: string | null
  model: string | null
  request_count: number
  input_tokens: number
  output_tokens: number
  total_cost_cny: number
}

export interface CostSummary {
  total_requests: number
  total_input_tokens: number
  total_output_tokens: number
  total_cost_cny: number
  by_provider: { provider_id: string; provider_name: string; request_count: number; total_cost_cny: number }[]
  by_model: { model: string; request_count: number; total_cost_cny: number }[]
}

export interface DateRange {
  start: string
  end: string
}

export interface LogFilter {
  limit?: number
  offset?: number
  provider_id?: string
  model?: string
}

export const getCostSummary = (range: DateRange) => invoke<CostSummary>('get_cost_summary', { range })
export const getRequestLogs = (filter: LogFilter) => invoke<RequestLog[]>('get_request_logs', { filter })
export const getDailyUsage = (range: DateRange) => invoke<DailyUsage[]>('get_daily_usage', { range })

File: src-tauri/src/commands/cost.rs
use tauri::State;
use crate::db::{models::RequestLog, models::DailyUsage, models::CostSummary, queries, DbState};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct DateRange {
    pub start: String,
    pub end: String,
}

#[derive(Debug, Deserialize)]
pub struct LogFilter {
    pub limit: Option<i64>,
    pub offset: Option<i64>,
    pub provider_id: Option<String>,
    pub model: Option<String>,
}

#[tauri::command]
pub fn get_cost_summary(db: State<DbState>, range: DateRange) -> Result<CostSummary, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;

    let daily = queries::get_daily_usage(&conn, &range.start, &range.end)
        .map_err(|e| e.to_string())?;

    let mut total_requests: i64 = 0;
    let mut total_input: i64 = 0;
    let mut total_output: i64 = 0;
    let mut total_cost: f64 = 0.0;

    for d in &daily {
        total_requests += d.request_count;
        total_input += d.input_tokens;
        total_output += d.output_tokens;
        total_cost += d.total_cost_cny;
    }

    Ok(CostSummary {
        total_requests,
        total_input_tokens: total_input,
        total_output_tokens: total_output,
        total_cost_cny: total_cost,
        by_provider: vec![],
        by_model: vec![],
    })
}

#[tauri::command]
pub fn get_request_logs(db: State<DbState>, filter: LogFilter) -> Result<Vec<RequestLog>, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    let limit = filter.limit.unwrap_or(50);
    let offset = filter.offset.unwrap_or(0);
    queries::get_request_logs(&conn, limit, offset).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_daily_usage(db: State<DbState>, range: DateRange) -> Result<Vec<DailyUsage>, String> {
    let conn = db.0.lock().map_err(|e| e.to_string())?;
    queries::get_daily_usage(&conn, &range.start, &range.end).map_err(|e| e.to_string())
}
